### PR TITLE
feat(quantic): Creation of a quanticQuickviewContent component

### PR DESCRIPTION
**Pull request by SimonMilord**
[SFINT-4936](https://coveord.atlassian.net/browse/SFINT-4936)

**Description**

This is a fix for the [youtube iframe bug](https://coveord.atlassian.net/browse/SFINT-4769)
Creation of a quanticQuickviewContent component in LWC. We decided to render only the youtube iframe in the content modal and not the _excerpts_ @obonneau let us know if you feel the excerpts for youtube content is necessary. We can add it later, it is just slightly more work to get it formatted properly.
 
The new quickview component handles the youtube iframes properly while still handling other types of documents the same way it had before. The use of render method provides us with the flexibility to add more templates in the future, should other similar exceptions arise.

**DEMO:**

https://user-images.githubusercontent.com/73316533/227291707-30b85045-7db3-42d4-97dc-e4a8a6f3fc9a.mov

[SFINT-4936]: https://coveord.atlassian.net/browse/SFINT-4936?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.html
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.html:1]: It would really help the reviewers to only see the lines that actually changed without the lines that changed duo to the formatting, especially that we decided to address the formatting of the HTML files in a different  ticket [SFINT-4935](https://coveord.atlassian.net/browse/SFINT-4935)

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.css
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.css:8]: These two lines can be replaced by the following SLDS CSS classes: `slds-grid` and `slds-grid_vertical`
https://www.lightningdesignsystem.com/utilities/grid/#Top-to-Bottom

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.html
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.html:89]: pretty sure we don't need this.
```suggestion
```

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js:123]: Why are you adding this typings?  does it show you a red underline when you try to access `error.message` ?

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js:187]: same remark for all the typings, I don't think they are necessary.

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.css
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.css:9]: Also you can add the styles of this class to the already existing class `quickview__content-container` since both are used in the same element instead of creating a new class.

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.html
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.html:86]: Are these classes really affecting the look of the `c-quantic-quickview-content` component?

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewYoutube.html
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewYoutube.html:21]: Do we need all this directives to make it work?
Let's keep only what we need.

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.js
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.js:19]: a better description would be appreciated.

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.js
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.js:11]: The js doc to document the component would be nice! take the other components as a reference.
Also this I think this component can be tagged as `@internal` since it's really an internal component and we don't really want it to appear in the Quantic documentation.

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewYoutube.css
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewYoutube.css:2]: why exactly do we need the result actions styles here?

**Review by mmitiche — CHANGES_REQUESTED** [review]
Great job finding the solution 👌

I'm mainly requesting changes to address some refinement  and  toremove the new exposed properties of the QuanticResultQuickview component, cause public properties are quite dangerous. They are known for being the number one reason causing breaking changes x)

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js:117]: it showed a warning, but not an error, I can remove those :P

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js]: you are right, I changed it to have them in the childcomponent. We dont need them here

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.css
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.css:10]: this was following a discussion with @erocheleau we had agreed on 90% for the modal content
otherwise it gives us something like this: 
image

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.html
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.html]: yes they are needed, for example if I remove the slds-modal__content it removes the background for the modal where the content is

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewYoutube.html
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewYoutube.html:21]: I removed the sandbox, it is not needed. However, I originally put the permissions because I was getting errors in the console. I was able to avoid console errors after removing the sandbox by allowing youtube as a trusted site in the CSP Trusted Sites settings of the scratch org.

https://developer.salesforce.com/docs/atlas.en-us.lightning.meta/lightning/csp_trusted_sites.htm

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewYoutube.css
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewYoutube.css:2]: good catch will remove

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.js
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.js:26]: oh you are right, it seems to work even without 😅

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewDefault.html
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewDefault.html:17]: yeah just read a bit on this, it seems better to include in css rather than the iframe tag properties.

More flexible, better for accessibility and avoid certain issues with putting them directly in the tag (ie performance loading issues, display issues)

File: packages/quantic/cypress/e2e/default-2/result-quickview/result-quickview.cypress.ts
Comment (SimonMilord) [inline packages/quantic/cypress/e2e/default-2/result-quickview/result-quickview.cypress.ts:45]: we want the spinner to display when loading

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.css
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.css:20]: I refactored as much as I could while making sure the modal content looks good

**Review by mmitiche — CHANGES_REQUESTED** [review]
As mentioned in the comments, the `footer` slot needs to remain in the QuanticResultQuickview component, otherwise it would be a breaking change in already existing implementations using this this component.

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.js
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.js:41]: makes sense!

File: packages/quantic/cypress/e2e/default-2/result-quickview/result-quickview-selectors.ts
Comment (SimonMilord) [reply packages/quantic/cypress/e2e/default-2/result-quickview/result-quickview-selectors.ts:40]: yes this also works

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js:217]: you should have something like: 
```
this.addEventListener('renderfacet', this.handleLoadingStateChange);
```
in the connected callback of this component in order to properly listen to the event.

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.js
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.js:38]: It would be nice if we can add a unit test that confirms that the QuanticQuickviewContent component is properly dispatching this event.

Take a look at this test in the QuanticResultAction component, it can give you an idea on how we can achieve such a test: https://github.com/coveo/ui-kit/blob/758f2f0f0f5a1d6dbf94c279563c04d999cc699a/packages/quantic/force-app/main/default/lwc/quanticResultAction/__tests__/quanticResultAction.test.js#L232-L247

Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.js:38]: Also since we always send the value `false` in the key `isLoading` of the payload, we can reduce complexity by completely omitting this payload.

Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.js]: yeah I removed the payload and changed the event, thats a good point

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js:128]: This should not be included in the  `if (this.isResultAction)` condition block.

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js:231]: Just to play it safe it would be better to stop the propagation of the event after handling it.
```suggestion
  handleLoadingStateChange(event) {
    event.stopPropagation();
    this._isLoading = false;
  }
```

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewYoutube.html
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewYoutube.html:4]: I don't see the function `stopPropagation` defined in the JS file, we need to define it.

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js:128]: I must have made a mis-type here, it was meant to be under the condition

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewYoutube.html
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewYoutube.html:1]: Now that I see the full component, I really wonder if having two templates `quanticQuickviewDefault.html` and `quanticQuickviewYoutube.html` is the best solution .
Cause After all the difference between the two templates is mainly just the attribute `src` passed to the iframe.
We can create a getter that returns the right src instead, this would simplify the component by removing one of the template and 2 of the CSS files.

We usually prefer rendering multiple template when the difference is huge between the templates.
In that case the difference is very subtle.

What do you think?

**Review by mmitiche — APPROVED** [review]
Nice job!
Raised one last concern about the usage of multiple templates, I'll leave that to you and the second reviewer.

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewYoutube.html
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewYoutube.html:1]: I think this is a valid point. However, we originally had decided to use multiple templates because of the following reasons:

- the current youtube template might change in the future where we might end up adding excerpts later on which will make the template different than the default one
- the use of the templates with the render function makes it more flexible in case we have to handle new types of results in the future which would need their own template.

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.css
Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.css:4]: It's based off the existing quickview, it takes maximum 90% of the content.

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.js
Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.js]: Yea at some point they made @track decoration obsolete I think.
Feels weird to rely on it being automagic instead of specifying it though :)

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/__tests__/quanticQuickviewContent.test.js
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/__tests__/quanticQuickviewContent.test.js:108]: I would put this expect before the listener one no? 
If the iframe is null it will for sure never call the listener, but the iframe could be non-null and never call the listener?

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.js
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.js:46]: Interesting, wouldn't have thought to use that as the check, but I think it works well :+1:

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewDefault.css
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewDefault.css:1]: Do you really need to make it's own css file? Does it not load the `quanticQuickviewContent.css` for both templates if you don't make those specific css files?

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewYoutube.html
Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewYoutube.html:1]: I think I would keep both templates too, but if possible remove the 2 specific CSS files?

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js
Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js:226]: Yea, also, I think I would call it `doneloading`.

**Review by erocheleau — COMMENTED** [review]
Looking good, small comments otherwise this is really close to :go:

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.js
Comment (nathanlb) [inline packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.js:17]: While this is actually pretty cool and the doc generator could be configured to use a tag like this instead, currently the way to omit a component from the reference doc JSON is to set `@category Internal`

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.js
Comment (nathanlb) [inline packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.js:14]:
```suggestion
 * @category Internal
```

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewDefault.css
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewDefault.css:1]: I do unfortunately. LWC doesnt support the import of CSS files in js and I didn't want to add inline styles. Salesforce links the CSS files with corresponding HTML files. Its seems the best solution for now is 1 shared css file with imports in name-specific css files. Unless you know another way?

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/__tests__/quanticQuickviewContent.test.js
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/__tests__/quanticQuickviewContent.test.js:48]: good point, ill add JSON files

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewDefault.css
Comment (nathanlb) [reply packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewDefault.css:1]: I'm also surprised that each .html template needs its own .css 🤔 but I haven't checked so if you say it's necessary so maybe it is.

**Review by nathanlb — APPROVED** [review]
Maybe @erocheleau wants to finish his review before you merge but LGTM

File: packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.js
Comment (nathanlb) [reply packages/quantic/force-app/main/default/lwc/quanticQuickviewContent/quanticQuickviewContent.js:17]: You can remove this now that the @category has been updated.

**Review by erocheleau — APPROVED** [review]
Looks good, just to remove the `@internal` in the jsdoc since you have the `@category Internal` :)

**Pull request merged**
